### ISSUE TITLE
Speed up `math.ndim` for `tensorflow`

### ISSUE THREAD

**Issue by dwierichs** [PennyLaneAI/pennylane#4691]
### Feature details

It seems that `qml.math.ndim(tensor)` is rather slow for  `tensorflow`, compared to an (equivalent, I think) `len(math.shape(tensor))` or `len(tensor.shape)` call. See timings below (`array` is a custom function that simply casts to a `tf.Variable`).
As we can see, `len(tensor.shape)` is the fastest with about $0.1 \mu s$, `len(math.shape(state))` consistently lies around $3 \mu s$, whereas the `math.ndim` call takes more than $60 \mu s$, of which little over $40\mu s$ are due to the Tensorflow implementation of `tf.experimental.numpy.ndim` itself.

We will have to see whether there are cases in which `len(tensor.shape)` fails that are covered by `tf.experimental.numpy.ndim`, but even if so, we should revert the fallback logic of `_ndim_tf` in `single_dispatch.py` to first try the `len(tensor.shape)` call and fall back to `ndim`.

![image](https://github.com/PennyLaneAI/pennylane/assets/20772922/7f7dc7ff-3739-41c1-9ea6-97963ad3213d)

### Implementation

Change `_ndim_tf` from

```python
def _ndim_tf(tensor):
    try:
        ndim = _i("tf").experimental.numpy.ndim(tensor)
        if ndim is None:
            return len(tensor.shape)
        return ndim
    except AttributeError:
        return len(tensor.shape) 
```
to

```python
def _ndim_tf(tensor):
    try:
        return len(tensor.shape)
    except AttributeError:
        return _i("tf").experimental.numpy.ndim(tensor)
```

This should deliver a speedup factor of about 300 to standard `ndim` on standard tensorflow objects.
As the new device frequently call `math.ndim` on the state vector, this is a relevant speedup that suppresses the cost of this basic evaluation to "negligible".

It remains to be seen whether slowdowns because of the reverted try-except logic are relevant to us, in particular when tracing.

### How important would you say this feature is?

1: Not important. Would be nice to have.

**Comment by dwierichs** [comment]
Thanks for the input @timmysilv! It's definitely hard to think ahead about all use cases here. I realized in the meantime that we can also simply revert to `len(shape)` in places where we are sure it makes sense/works, instead of changing `ndim`. :sweat_smile:

**Comment by timmysilv** [comment]
note that I just merged #4847, changing `ndim` for `tf` to simply return `len(tensor.shape)`. If it's an abstract tensor with the shape specified as None in a `TensorSpec`, it'll fail in the same way as it did before and get handled by `Operator._check_batching`.

Can we call this issue complete? I just checked locally, and `len(tensor.shape)` is typically around 0.2 $\mu s$, whereas `qml.math.ndim(tensor)` is around 1.5 $\mu s$, but 1.5 is still significantly better than the ~50 $\mu s$ you were seeing in these benchmarks
